File: openbis/source/java/ch/systemsx/cisd/openbis/public/resources/api/v3/dto/search/SampleSearchCriteria.js
/**
 * @author pkupczyk
 */
define([ "require", "stjs", "dto/search/AbstractEntitySearchCriteria", "dto/search/SampleSearchRelation", "dto/search/SpaceSearchCriteria", 
		"dto/search/ProjectSearchCriteria", "dto/search/NoProjectSearchCriteria", "dto/search/ExperimentSearchCriteria",
		"dto/search/NoExperimentSearchCriteria", "dto/search/NoSampleContainerSearchCriteria", "dto/search/SearchOperator" ], function(require, stjs, AbstractEntitySearchCriteria,
		SampleSearchRelation, SpaceSearchCriteria, ExperimentSearchCriteria, NoExperimentSearchCriteria, NoSampleContainerSearchCriteria, SearchOperator) {

	var SampleSearchCriteria = function(relation) {
		AbstractEntitySearchCriteria.call(this);
		this.relation = relation ? relation : SampleSearchRelation.SAMPLE;
	};
	stjs.extend(SampleSearchCriteria, AbstractEntitySearchCriteria, [ AbstractEntitySearchCriteria ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.SampleSearchCriteria';
		constructor.serialVersionUID = 1;
		prototype.relation = null;
		prototype.withSpace = function() {
			return this.addCriteria(new SpaceSearchCriteria());
		};
		prototype.withProject = function() {
			return this.addCriteria(new ProjectSearchCriteria());
		};
		prototype.withoutProject = function() {
			return this.addCriteria(new NoProjectSearchCriteria());
		};
		prototype.withExperiment = function() {
			return this.addCriteria(new ExperimentSearchCriteria());
		};
		prototype.withoutExperiment = function() {
			return this.addCriteria(new NoExperimentSearchCriteria());
		};
		prototype.withParents = function() {
			return this.addCriteria(new SampleParentsSearchCriteria());
		};
		prototype.withChildren = function() {
			return this.addCriteria(new SampleChildrenSearchCriteria());
		};
		prototype.withContainer = function() {
			return this.addCriteria(new SampleContainerSearchCriteria());
		};
		prototype.withoutContainer = function() {
			return this.addCriteria(new NoSampleContainerSearchCriteria());
		};
		prototype.withOrOperator = function() {
			return this.withOperator(SearchOperator.OR);
		};
		prototype.withAndOperator = function() {
			return this.withOperator(SearchOperator.AND);
		};
		prototype.getRelation = function() {
			return this.relation;
		};
		prototype.createBuilder = function() {
			var builder = AbstractEntitySearchCriteria.prototype.createBuilder.call(this);
			builder.setName(this.relation.name());
			return builder;
		};
	}, {
		relation : {
			name : "Enum",
			arguments : [ "SampleSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriteria" ]
		}
	});

	var SampleParentsSearchCriteria = function() {
		SampleSearchCriteria.call(this, SampleSearchRelation.PARENTS);
	};
	stjs.extend(SampleParentsSearchCriteria, SampleSearchCriteria, [ SampleSearchCriteria ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.SampleParentsSearchCriteria';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "SampleSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriteria" ]
		}
	});

	var SampleChildrenSearchCriteria = function() {
		SampleSearchCriteria.call(this, SampleSearchRelation.CHILDREN);
	};
	stjs.extend(SampleChildrenSearchCriteria, SampleSearchCriteria, [ SampleSearchCriteria ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.SampleChildrenSearchCriteria';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "SampleSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriteria" ]
		}
	});

	var SampleContainerSearchCriteria = function() {
		SampleSearchCriteria.call(this, SampleSearchRelation.CONTAINER);
	};
	stjs.extend(SampleContainerSearchCriteria, SampleSearchCriteria, [ SampleSearchCriteria ], function(constructor, prototype) {
		prototype['@type'] = 'dto.search.SampleContainerSearchCriteria';
		constructor.serialVersionUID = 1;
	}, {
		relation : {
			name : "Enum",
			arguments : [ "SampleSearchRelation" ]
		},
		operator : {
			name : "Enum",
			arguments : [ "SearchOperator" ]
		},
		criteria : {
			name : "Collection",
			arguments : [ "ISearchCriteria" ]
		}
	});

	return SampleSearchCriteria;
})

File: openbis_api/source/java/ch/ethz/sis/openbis/generic/shared/api/v3/dto/search/SampleSearchCriteria.java
/*
 * Copyright 2014 ETH Zuerich, CISD
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.ethz.sis.openbis.generic.shared.api.v3.dto.search;

import ch.ethz.sis.openbis.generic.shared.api.v3.dto.id.sample.ISampleId;
import ch.systemsx.cisd.base.annotation.JsonObject;

/**
 * @author pkupczyk
 */
@JsonObject("dto.search.SampleSearchCriteria")
public class SampleSearchCriteria extends AbstractEntitySearchCriteria<ISampleId>
{

    private static final long serialVersionUID = 1L;

    private SampleSearchRelation relation;

    public SampleSearchCriteria()
    {
        this(SampleSearchRelation.SAMPLE);
    }

    SampleSearchCriteria(SampleSearchRelation relation)
    {
        this.relation = relation;
    }

    public SpaceSearchCriteria withSpace()
    {
        return with(new SpaceSearchCriteria());
    }

    public ProjectSearchCriteria withProject()
    {
        return with(new ProjectSearchCriteria());
    }

    public SampleSearchCriteria withoutProject()
    {
        with(new NoProjectSearchCriteria());
        return this;
    }
    
    public ExperimentSearchCriteria withExperiment()
    {
        return with(new ExperimentSearchCriteria());
    }

    public SampleSearchCriteria withoutExperiment()
    {
        with(new NoExperimentSearchCriteria());
        return this;
    }

    public SampleParentsSearchCriteria withParents()
    {
        return with(new SampleParentsSearchCriteria());
    }

    public SampleChildrenSearchCriteria withChildren()
    {
        return with(new SampleChildrenSearchCriteria());
    }

    public SampleContainerSearchCriteria withContainer()
    {
        return with(new SampleContainerSearchCriteria());
    }

    public SampleSearchCriteria withoutContainer()
    {
        with(new NoSampleContainerSearchCriteria());
        return this;
    }

    public SampleSearchCriteria withOrOperator()
    {
        return (SampleSearchCriteria) withOperator(SearchOperator.OR);
    }

    public SampleSearchCriteria withAndOperator()
    {
        return (SampleSearchCriteria) withOperator(SearchOperator.AND);
    }

    public SampleSearchRelation getRelation()
    {
        return relation;
    }

    @Override
    protected SearchCriteriaToStringBuilder createBuilder()
    {
        SearchCriteriaToStringBuilder builder = super.createBuilder();
        builder.setName(relation.name());
        return builder;
    }

}
